src/atomic.cr

```diff
@@ -157,7 +157,12 @@ struct Atomic(T)
   # atomic.get      # => 10
   # ```
   def swap(value : T)
-    Ops.atomicrmw(:xchg, pointerof(@value), value, :sequentially_consistent, false)
+    {% if T.union? && T.union_types.all? { |t| t == Nil || t < Reference } || T < Reference %}
```

Comment:
Look at the `compare_and_set` method above for how to implement this. It uses different cases for nillable references and references. Currently this wont work for nilable references.

Reply:
It does work and I added a test specifically for that case. The difference from ```compare_and_set``` has to do with the  return value of the method, no the type of ```T```.

Reply:
Oh, my bad.

It looks to me then like `compare_and_set` should work without the special case for nilables as `Pointer(Void).null.as(String?)` == `nil`.

Reply:
I tried what you suggested and it appears to work but perhaps whoever implemented it knows of some edge case we do not.
Anyway, that is a matter for another issue/PR.

Reply:
@Exilor yup, which is why I approved this :)